### ISSUE TITLE
Predefined object notation for Date

### ISSUE THREAD

**Issue by roccx** [JamitLabs/SimpleAPISpec#1]
It is very common to use different representations of dates in APIs. By defining a predefined set of possible representations we could simplify the generic implementation of a client and also reduce error sources happening by mapping dates.

I would suggest predefining a notation for the Date object, which will at least provide an interface to allow customizations and also enable the possibility to let the API decide which representation will be sent without the requirement to migrate the client implementation for noncustom representations.

**Comment by Dschee** [comment]
Good idea. I find, we should even go so far to define a default date format in case an API doesn't explicitly define a different one, like we already do with the pagination page size (which is 25 by default if not specified otherwise). I would recommend to use [ISO 8601:2004](https://www.iso.org/standard/40874.html) for that.

Meaning we should specify `Date`, `DateTime` and `Time` default formats based on 8601 like so:
Date format: `YYYY-MM-DD`
Time format: `hh:mm:ss`
DateTime format: `YYYY-MM-DDThh:mm:ssZ`

**Comment by roccx** [comment]
It would be really good to use a default format based on a specification.

So we can say that if APIs want to customize their Date format, they also must maintain it in their client implementation?

Or maybe we can also send the format in the body or would this be too much overhead?
Something like:
```
{
    "typeName": "Date",
    "format": "ISO-8601",
    "value": "..."
}
```

As an alternative we could reserve the keywords `Date`, `DateTime` and `Time` for the `typeName` property to always match ISO 8601.

**Comment by Dschee** [comment]
I'm against sending the format in the body – it's not what I understand as a "simple" API spec, it's more like a "rich" API spec then which is a non-goal for this project. I think the simplest thing for solving the date problem is to define a default format which works in nearly all cases and if another one is required, the API has to define it's own, whatever that might be (including your suggested solution).

**Comment by Dschee** [comment]
Please also note that we are not defining any `typeName` keywords and if we would want to define dates somewhere, then that would be the type of properties, the `typeName` is for the class name though, so there's a mismatch or misunderstanding there. I think we should just document, that fields which hold dates should use the format of one of the three types. Whatever best fits the naming convention of the API programmers should be used, but we should explain the difference between the three and an API can either directly use the same names or they can explain what their naming means regarding the date format we define here.

**Comment by roccx** [comment]
I see that this approach will introduce too much complexity so we can use `ISO 8601` as default and do not provide further specifications.

My problem with the `typeName` is that JSON doesn't provide any Date types. So it would be parsed as a string value and wouldn't be validatable as correct Date body. My idea was to let the spec itself deliver enough information to enable date field detection and format validations.

**Comment by Dschee** [comment]
I see what you mean. But when we start improving JSON we should also consider other common types like geographic location (which is currently two fields of Doubles) and many other things. The current solution would be that the API documentation has to specify this string key to be a `Date` for example, then both the server and client can specify the field to be a `Date` and the fact that it's actually a String in the JSON would be an implementation detail.

Think of it this way: When there would be a successor for JSON which incorporates a specific format for dates and other useful types, then nothing would change, the API would still specify it to be a `Date` and the library would just implement the reading of the JSON-successor differently. If we had defined a format ourselves, then we would have redundant information in that JSON-successor. Although the spec currently only supports JSON, I don't think that has to be the case forever (there's already some work done in this direction, like [Transit](http://blog.cognitect.com/blog/2014/7/22/transit)).

**Comment by Dschee** [comment]
By the way, we actually could put automatic `Date`, `DateTime` and `Time` detection in place for client/server libraries if we wanted, the above suggested format is already quite usable for that and it's standardized by ISO 8601 already, so we don't have to teach anybody about it. But even with automatic detection in place, I think having the server and client developers state the expected type of a field explicitly is always better then doing something implicitly based on the data.
